### ISSUE TITLE
gRPC Support

### ISSUE THREAD

**Issue by luigibarbato** [Orange-OpenSource/hurl#3411]
### Problem to solve
I have been using Hurl extensively so far to test internal REST services. Now we are switching to gRPC and since I have not read any mention about it in the Hurl docs I am afraid that it is not supported and we will have to migrate to another tool. I am asking if there is currently support for gRPC and if so is there any idea of supporting it. Thank you very much for the great work to all of you on the Hurl team.

**Comment by njuptlzf** [comment]
like grpcurl

**Comment by jwadolowski** [comment]
@jcamiel [`grpcurl`](https://github.com/fullstorydev/grpcurl) is a generic CLI gRPC client and it doesn't come with any testing/response validation capabilities.

**Comment by sify21** [comment]
@jcamiel Adding some other resources for referece
- https://github.com/ktr0731/evans
  It supports `REPL` and `CLI` mode. Written in Go. (unmaintained)
- https://github.com/grpc-ecosystem/polyglot
  A grpc command-line client written in Java. (unmaintained)
- https://learn.microsoft.com/en-us/aspnet/core/grpc/test-tools
  Documentation from ASP.NET showing how to test grpc services useing grpcurl
- https://github.com/grpc/grpc/blob/master/test/cpp/util/grpc_cli.cc
  CLI tool from grpc's main repo, written in C. (There is also a [doc](https://github.com/grpc/grpc/blob/master/doc/command_line_tool.md) for it)

**Comment by jcamiel** [comment]
@sify21 thanks a lot very helpful

**Comment by talgat-ruby** [comment]
I use [httpyac](https://httpyac.github.io/)

**Comment by LeoBorai** [comment]
Hi guys! Any news an this? I would be glad to use Hurl for testing my gRPC servers.

I'm happy to contribute on this if possible.

**Comment by ppaulweber** [comment]
@jcamiel any updates on this topic? Or did someone already started some initial effort? Happy to help, support, or start efforts as well in this regard.

**Comment by jcamiel** [comment]
Hi @ppaulweber 

Yes we've begun to seriously work on gRPC support for Hurl. The main works happens for the moment in this repo <https://github.com/jcamiel/grpc>. We want to propose a "beta" version in Hurl this autumn. I say "beta" because we're not at all gRPC experts so there could be a lot of errors/misconceptions etc... At a point, we'll make a call-to-test for people interested in to see if we're in a good shape or not.

What we see in a first version:

- support only proto3 and only binary FileDescriptorSet (`--protoset` grpcurl option)
- unary request/response to fit in Hurl model (we plan to expand later this model to have streaming, servers responses etc... but not only for gRPC also SSE for instance)
- no reflection

On the syntax side, it will be something like this:

- Sending a request

~~~hurl
POST http://localhost:50051/helloworld.Greeter/SayHello
```grpc
{
  "name": "bob"
}
```
HTTP 200
~~~

run with `hurl --grpc-protoset helloworld.protoset hello.hurl`

- Sending a request with options:
~~~hurl
POST http://localhost:50051/helloworld.Greeter/SayHello
[Options]
grpc-protoset: helloworld.protoset
grpc-protoset: error_details.protoset
```grpc
{
  "name": "bob"
}
```
HTTP 200
~~~

- Assertions (to be refined)

~~~hurl
POST http://localhost:50051/helloworld.Greeter/SayHello
```grpc
{
  "name": "bob"
}
```
HTTP 200
[Asserts]
# Access to grpc-status response headers
grpc "status" == 0
grpc "message" == "foo"
jsonpath "$.message" == "Hello, bob"
~~~

Regarding the response, we should decide for instance if we automatically show default value even if they're absent from the wire (like grpcurl `-emit-defaults`) or do an option for this. Do we need to get access to the "real" (on-the-wire) binary response?

Another question this request:

~~~hurl
POST http://localhost:50051/helloworld.Greeter/SayHello
```grpc
{
  "name": "bob"
}
```
HTTP 200
~~~

could be OK because we receive an HTTP 200, but semantically it's failing beacuse the `grpc-status` header is different from OK. Do we want to fail automatically in this case (with an option like `--grpc-fail`)?

Regarding the request, having a request body with ` ```grpc ` will automatically enable `--http2-prior-knowledge` Hurl option (to force HTTP/2 without HTTP/1 negotiation)/

All the Hurl options (like `--proxy`, `--resolve` etc...) will work out of the box.

On the code side, we're trying _not to_ add any dependencies. I don't know if it's too ambitious, but, fo my part, I don't want to add a lot of crates to support this new feature. The community tests will show if our approach is feasible or not.

To sum up: for the moment, nothing is ready to be test, we're target this autumn for a beta version. 

In any case, any feedback is welcome (regarding the use case, the syntax etc...)

**Comment by nu11ptr** [comment]
Glad to see this. 

Might I be so bold as to suggest while carving out a syntax for gRPC that you consider either adding, or at least planning for, how to add [ConnectRPC](https://connectrpc.com) support at the same time? If you aren't aware, ConnectRPC is similar to gRPC (they both use proto files and protobuf), and is considered by many to be a simpler, more streamlined gRPC (by allowing HTTP 1.1 and JSON, in addition to HTTP/2 and Protobuf). It is gaining popularity (Anthropic has recently added crates for [first class Rust support](https://github.com/connectrpc/connect-rust)), is gRPC compatible, and likely would share 90%+ of the same syntax. Even if deemed out of scope, it would be nice to at least be aware of it so if the choice is made to add it later it fits nicely into whatever syntax scheme you pick for gRPC.
